packages/SwingSet/src/controller/controller.js

```diff
@@ -164,314 +168,400 @@ export async function makeSwingsetController(
-    bundleHandler,
-  };
-  const kernelRuntimeOptions = {
-    verbose,
-    warehousePolicy,
-    overrideVatManagerOptions,
-  };
-
-  /** @type { ReturnType<typeof import('../kernel/kernel.js').default> } */
-  const kernel = buildKernel(
-    kernelEndowments,
-    deviceEndowments,
-    kernelRuntimeOptions,
-  );
-
-  await kernel.start();
-
   /**
-   * Validate and install a code bundle.
+   * Capture an extended process in the slog, writing an entry with `type`
+   * $startLabel and then later (if the function returns successfully or calls
+   * the finish callback provided to it) another entry with `type` $endLabel and
+   * a `seconds` property valued with the total elapsed duration in seconds.
    *
-   * @param {EndoZipBase64Bundle} bundle
-   * @param {BundleID} [allegedBundleID]
-   * @returns {Promise<BundleID>}
+   * @template T
+   * @template {unknown[]} A
+   * @param {readonly [startLabel: string, endLabel: string]} labels
+   * @param {SlogDurationProps} startProps for both slog entries
+   * @param {(finish: (extraProps?: SlogDurationProps) => void, ...args: A) => (T | Promise<T>)} fn
+   * @param {unknown[] & A} args
+   * @returns {Promise<T>}
    */
-  async function validateAndInstallBundle(bundle, allegedBundleID = undefined) {
-    // TODO The following assertion may be removed when checkBundle subsumes
-    // the responsibility to verify the permanence of a bundle's properties.
-    // https://github.com/endojs/endo/issues/1106
-    mustMatch(bundle, endoZipBase64Sha512Shape);
-    await checkBundle(bundle, computeSha512, allegedBundleID);
-    const { endoZipBase64Sha512 } = bundle;
-    assert.typeof(endoZipBase64Sha512, 'string');
-    const bundleID = `b1-${endoZipBase64Sha512}`;
-    if (allegedBundleID !== undefined) {
-      bundleID === allegedBundleID ||
-        Fail`alleged bundleID ${allegedBundleID} does not match actual ${bundleID}`;
+  const slogDuration = async (labels, startProps, fn, ...args) => {
+    const [startLabel, endLabel] = labels;
+    const props = { ...startProps };
+    if (hasOwn(props, 'type') || hasOwn(props, 'seconds')) {
+      console.error('startProps must not include "type" or "seconds"');
+      delete props.type;
+      delete props.seconds;
     }
-    await kernel.installBundle(bundleID, bundle);
-    return bundleID;
-  }
-
-  // the kernel won't leak our objects into the Vats, we must do
-  // the same in this wrapper
-  const controller = harden({
-    log(str) {
-      kernel.log(str);
-    },
-
-    writeSlogObject,
+    let cause;
+    let finished = false;
+    /** @type {(extraProps?: SlogDurationProps) => void} */
+    const finish = extraProps => {
+      const seconds = (performance.now() - t0) / 1000;
+      if (finished) {
+        // `finish` should only be called once.
+        // Log a stack-bearing error instance, but throw something more opaque.
+        const msg = `slog event ${startLabel} already finished; ignoring props ${q(extraProps || {})}`;
```

Comment:
I think adding the start props would help identifying any erroneous finish slog event.

```suggestion
        const msg = `slog event ${startLabel} (${q(startProps || {})}) already finished; ignoring props ${q(extraProps || {})}`;
```